wallet-new/test/state-machine/Wallet.hs

```diff
@@ -112,21 +114,36 @@ instance ToExpr Microsecond where
 deriving instance ToExpr (Model Concrete)
 
 initModel :: Model r
-initModel = Model []
+initModel = Model [] False
 
--- TODO: make smarter preconditions. We currently don't use it as we
--- are weighting more closely distribution with `generator` function.
 -- If you need more fine grained distribution, use preconditions
 preconditions :: Model Symbolic -> Action Symbolic -> Logic
 preconditions _ ResetWalletA      = Top
-preconditions _ (CreateWalletA _) = Top
-preconditions _ GetWalletsA       = Top
+-- NOTE: with this mechanism we are forcing ResetWalletA to be first action
+-- in an action list generated. We need this because we are reusing same
+-- db through all quickcheck runs - so we have to ensure ResetActionA is run
+-- before any other action. Implementation by precondition is good enough and
+-- generator won't need to try too many options until it reaches ResetActionA.
+-- An alternative solution would be to modify `forAllCommands` in this way:
+--
+-- ```
+-- forAllCommands' sm n action = forAllCommands sm n $ \cmds -> action $ Commands [Command ResetWalletA mempty] <> cmds
+-- ```
+--
+-- The above would work a bit more performant (as we don't have failing preconditions)
+-- but we are hacking around to lib internals which is not so idiomatic.
```

Comment:
Another way would be to change the generator:

```haskell
generator (Model _ reset) | reset = pure ResetWalletA
                          | otherwise = ...
```

Reply:
Thanks for the review - I have added just what you have suggested https://github.com/input-output-hk/cardano-sl/pull/3772/commits/ac1058b120af19f292176479cff290dee62bc8de (without noticing that you have proposed the same solution).

I will keep both precondition and generator (which is just optimization for what the precondition is doing) for now.